### PR TITLE
List nodes

### PR DESCRIPTION
**Pull request by simonswine**
Adds a public `ListNodes() (nodes []NodeState, err error)` method to list dataset agents.

### CODE REVIEW

File: client.go [hunk 1]
```diff
@@ -163,25 +167,35 @@ func (c Client) findIDInConfigurationsPayload(body io.ReadCloser, name string) (
 	return "", err
 }
 
+// ListNodes returns a list of dataset agent nodes from Flocker Control Service
+func (c *Client) ListNodes() (nodes []NodeState, err error) {
+	resp, err := c.get(c.getURL("state/nodes"))
+	if err != nil {
+		return
```
Comment (wallnerryan) [inline client.go]: Would typically like to see an actual error returned. 

The below returns no info

```
if err != nil {
        return
}
```

Preferred something like the below

```
if err != nil {
        return nil, err
}
```

or

```
if err != nil {
        return "", err
}
```

File: client.go [hunk 2]
```diff
@@ -163,25 +167,35 @@ func (c Client) findIDInConfigurationsPayload(body io.ReadCloser, name string) (
...
+	}
+	defer resp.Body.Close()
+
+	err = json.NewDecoder(resp.Body).Decode(&nodes)
+	if err != nil {
+		return
+	}
+	return
```
Comment (wallnerryan) [inline client.go]: shouldn't this return `nodes`  ?

Comment (simonswine) [inline client.go]: https://tour.golang.org/basics/7

File: client.go (hunk 1 above)
Comment (simonswine) [inline client.go]: https://tour.golang.org/basics/7

File: client.go [hunk 3]
```diff
@@ -163,25 +167,35 @@ func (c Client) findIDInConfigurationsPayload(body io.ReadCloser, name string) (
...
 
+// ListNodes returns a list of dataset agent nodes from Flocker Control Service
+func (c *Client) ListNodes() (nodes []NodeState, err error) {
+	resp, err := c.get(c.getURL("state/nodes"))
+	if err != nil {
+		return
+	}
+	defer resp.Body.Close()
+
+	err = json.NewDecoder(resp.Body).Decode(&nodes)
```
Comment (wallnerryan) [inline client.go:194]: not as familiar with go but with the pointer `&nodes` i was expecting a `var nodes []NodeState`, or does this reference a object differently?

File: client.go [hunk 4]
```diff
@@ -163,25 +167,35 @@ func (c Client) findIDInConfigurationsPayload(body io.ReadCloser, name string) (
...
+		return
+	}
+	return
+}
+
 // GetPrimaryUUID returns the UUID of the primary Flocker Control Service for
 // the given host.
 func (c Client) GetPrimaryUUID() (uuid string, err error) {
-	resp, err := c.get(c.getURL("state/nodes"))
+	states, err := c.ListNodes()
```
Comment (wallnerryan) [inline client.go:204]: prefer `states` to be `nodeStates` or `nodes` for readability. 

EDIT: realized this is sort of misused throughout the code, better to leave consistent, so ignore.

File: client.go (hunk 3 above)
Comment (simonswine) [inline client.go:194]: That's the same as the other ones as well, naked or named returns initialise the variable, just by the method headers: `nodes []NodeState, err error`

File: client.go [hunk 5]
```diff
@@ -163,25 +167,35 @@ func (c Client) findIDInConfigurationsPayload(body io.ReadCloser, name string) (
...
-				return s.UUID, nil
-			}
+	for _, s := range states {
+		if s.Host == c.clientIP {
+			return s.UUID, nil
 		}
-		return "", errStateNotFound
 	}
-	return "", err
+	return "", errStateNotFound
```
Comment (wallnerryan) [inline client.go]: `State not found by Dataset ID` is not the most correct error here, consider adding a new one that more accurately states what's going on like "Host not found, could not update primary uuid"

**Comment by wallnerryan** [comment]
@simonswine looks good, I overall like the addition of `ListNodes()`, a few comments. I will wait for your responses. I also havent had time to test this, but probably a good idea to add a test or two for ListNodes() that covers listing with nodes present and listing with nodes empty.

File: client.go (hunk 1 above)
Comment (wallrj) [inline client.go]: > https://tour.golang.org/basics/7

Well yeah, but 
- it doesn't follow the house style....that is, I don't see "naked return" statements anywhere else in this library, and 
- I (at least) find it really hard to read. And I'm not the only one:
  - https://github.com/kubernetes/kubernetes/pull/4263#discussion_r24597352
  - https://github.com/golang/go/wiki/CodeReviewComments#named-result-parameters

File: client.go (hunk 3 above)
Comment (wallnerryan) [inline client.go:194]: 👍

File: client.go (hunk 1 above)
Comment (wallnerryan) [inline client.go]: ah okay, naked return thanks for the pointer. But agreed, would like it to be consistent.

File: client.go (hunk 2 above)
Comment (wallnerryan) [inline client.go]: 👍

**Comment by simonswine** [comment]
@wallrj  @wallnerryan all your concerns should be fixed now

File: client.go [hunk 6]
```diff
@@ -163,25 +167,35 @@ func (c Client) findIDInConfigurationsPayload(body io.ReadCloser, name string) (
...
-				return s.UUID, nil
-			}
+	for _, s := range states {
+		if s.Host == c.clientIP {
+			return s.UUID, nil
 		}
-		return "", errStateNotFound
 	}
-	return "", err
+	return "", fmt.Errorf("No node not found with IP '%s'", c.clientIP)
```
Comment (wallrj) [inline client.go]: Typo "No node not found"
And I'd be tempted to include the the full list of nodes (in JSON format) in the error message to make it easier to debug misconfigurations....but I'll leave that up to you.

File: client_test.go [hunk 7]
```diff
@@ -102,26 +102,76 @@ func TestFindIDInConfigurationsPayload(t *testing.T) {
...
 		assert.Equal("/v1/state/nodes", r.URL.Path)
-		w.Write([]byte(fmt.Sprintf(`[{"host": "%s", "uuid": "%s"}]`, mockedHost, mockedPrimary)))
+		w.Write(data)
 	}))
 
 	host, port, err := getHostAndPortFromTestServer(ts)
 	assert.NoError(err)
 
 	c := newFlockerTestClient(host, port)
 	assert.NoError(err)
```
Comment (wallrj) [inline client_test.go]: Another one of these unnecessary assertions....but perhaps that's been removed in another branch.

Comment (simonswine) [inline client_test.go]: Fixed during the rebase

File: client.go (hunk 6 above)
Comment (simonswine) [inline client.go]: I am outputing  the list now (but not as a json object)

File: client.go [hunk 8]
```diff
@@ -178,25 +180,35 @@ func (c Client) findIDInConfigurationsPayload(body io.ReadCloser, name string) (
 	return "", err
 }
 
+// ListNodes returns a list of dataset agent nodes from Flocker Control Service
+func (c *Client) ListNodes() (nodes []NodeState, err error) {
+	resp, err := c.get(c.getURL("state/nodes"))
+	if err != nil {
+		return []NodeState{}, err
+	}
```
Comment (wallrj) [inline client.go:188]: We should return an error for HTTP response codes >=300 for consistency with the other methods and so that the JSON decoding doesn't blow up with the non-json response body that you get from 500 errors.

**Comment by wallrj** [comment]
Thanks @simonswine 

I'd like some HTTP response code checking in the new ListNodes method, but other than that it looks great.

**Comment by simonswine** [comment]
@wallrj  I agree and I have added that.

This should be probably done in the get method, but I did only refactor that part into a new method and did not want to change anything around the behaviour....
